### PR TITLE
Couchbase

### PR DESCRIPTION
**Pull request by julienstroheker**
## Waiting for review
* Adding CB deployment
* Have to add one Persistent Storage

### CODE REVIEW

File: incubator/couchbase/templates/couchbase-master-rc.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v1
+kind: ReplicationController
```
Comment (jpoon) [inline incubator/couchbase/templates/couchbase-master-rc.yaml]: Consider using a https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/ instead. It's a lot better with stateful scenarios and handles storage a lot better. (same comment for your worker).

File: incubator/couchbase/values.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+## couchbase image version
+## ref: https://hub.docker.com/r/library/couchbase/tags/
+##
+
+masterReplicaCount: 1
```
Comment (jpoon) [inline incubator/couchbase/values.yaml]: How does multi-master setup work? Does Couchbase auto-replicate the data across the masters?

File: incubator/couchbase/values.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+persistence:
+  enabled: true
+  ## If defined, volume.beta.kubernetes.io/storage-class: <storageClass>
+  ## Default: volume.alpha.kubernetes.io/storage-class: default
+  ##
+  storageClass: fast 
+  accessMode: ReadWriteOnce
+  size: 10Gi
+ 
+resources:
```
Comment (jpoon) [inline incubator/couchbase/values.yaml]: You don't use these anywhere.

File: incubator/couchbase/templates/couchbase-worker-rc.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    metadata:
+      labels:
+        app: couchbase-worker-pod
+    spec:
+      containers:
+      - name: couchbase-worker
+        image: couchbase/server:sandbox
+        env:
+          - name: TYPE
+            value: "WORKER"
```
Comment (jpoon) [inline incubator/couchbase/templates/couchbase-worker-rc.yaml]: You probably don't need the quotes around the `value`

File: incubator/couchbase/templates/couchbase-master-rc.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+metadata:
+  name: couchbase-master-rc
+spec:
```
Comment (jpoon) [inline incubator/couchbase/templates/couchbase-master-rc.yaml]: I think you're missing a volume mount here but I think you mentioned that you took that piece out for the time being cause it wasn't working?

File: incubator/couchbase/values.yaml (hunk 2 above)
Comment (julienstroheker) [inline incubator/couchbase/values.yaml]: With CB you dont have this concept of master and slave. You have one "master" which is the first node with the index, service etc... then, using the same image you have worker. BTW, the master is also a worker.
For the replication, you have to "rebalance" the data accross your cluster from the UI. I will document that part when the code will be clean.

Comment (jpoon) [inline incubator/couchbase/values.yaml]: In that case, does it make sense to have this as a variable? Parameters in `values.yaml` are meant to be configurable items.

Comment (julienstroheker) [inline incubator/couchbase/values.yaml]: Good point

**Comment by julienstroheker** [comment]
@jpoon Updated to StatefulSets, have to figure it out for the persitent storage now. Thanks for your advises !

**Comment by jpoon** [comment]
@julienstroheker I have some time later today. Let me know if you want to partner up and work on this.

**Comment by julienstroheker** [comment]
@jpoon should be ok now, with the Persistent Storage + your previous recommendations.
Thanks

File: incubator/couchbase/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+The following tables lists the configurable parameters of the Couchbase chart and their default values.
+
+|         Parameter            |             Description             |                         Default                          |
+|------------------------------|-------------------------------------|----------------------------------------------------------|
+| `workerReplicaCount`         | Number of replica worker            | `3`                                                      |
+| `persistence.enabled`        | Use a PVC to persist data           | `true`                                                   |
+| `persistence.storageClass`   | Storage class of backing PVC        | `fast`                                                   |
+| `persistence.accessMode`     | Use volume as ReadOnly or ReadWrite | `ReadWriteOnce`                                          |
+| `persistence.size`           | Size of data volume                 | `10Gi`                                                   |
+| `resources`                  | CPU/Memory resource requests/limits | Memory: `128Mi`, CPU: `100m`                             |
```
Comment (jpoon) [inline incubator/couchbase/README.md]: You don't look to be using `resources`.

File: incubator/couchbase/templates/couchbase-worker-statefulset.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  serviceName: {{ template "fullname" . }}
+  replicas: {{ .Values.workerReplicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "fullname" . }}
+        couchbaserole: "worker"
+    spec:
+      containers:
+      - name: couchbase-master
```
Comment (jpoon) [inline incubator/couchbase/templates/couchbase-worker-statefulset.yaml]: Did you mean `couchbase-slave` as the name?

File: incubator/couchbase/Chart.yaml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+name: couchbase
+version: 0.1.0
+description: Couchbase cluster with a master and 1 or more worker Replication Controller (RC).
```
Comment (jpoon) [inline incubator/couchbase/Chart.yaml]: "Couchbase cluster with a master and at least 1 worker"

File: incubator/couchbase/templates/couchbase-worker-statefulset.yaml (hunk 7 above)
Comment (julienstroheker) [inline incubator/couchbase/templates/couchbase-worker-statefulset.yaml]: `couchbase-worker` indeed

**Comment by julienstroheker** [comment]
```
==> Linting couchbase/
Lint OK

1 chart(s) linted, no failures
```

**Comment by jpoon** [comment]
Noooiceeee. Ship it.

**Pull request closed without merging**